awx/main/analytics/core.py

```
@@ -388,12 +389,23 @@ def ship(path):
+                cert_path, key_path = get_or_generate_client_certificate(rh_id, rh_secret)
+                if cert_path and key_path:
+                    logger.debug("Using certificate-based authentication for analytics upload")
+                    response = s.post(url, files=files, cert=(cert_path, key_path), verify=settings.INSIGHTS_CERT_PATH, headers=s.headers, timeout=(31, 31))
+                else:
+                    # Second attempt: OIDC authentication
+                    logger.debug("Certificate unavailable, falling back to OIDC authentication")
+                    client = OIDCClient(rh_id, rh_secret)
+                    response = client.make_request("POST", url, headers=s.headers, files=files, verify=settings.INSIGHTS_CERT_PATH, timeout=(31, 31))
+            except requests.RequestException as e:
+                # Final fallback: Basic authentication (until fully removed)
+                logger.error("Certificate and OIDC authentication failed (%s), trying basic auth", e)
```

Ruff (0.15.1)

[warning] Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)